File: 프로그래머스/unrated/161988. 연속 펄스 부분 수열의 합/연속 펄스 부분 수열의 합.js
function solution(sequence) {
  var answer = 0;

  const pseq = [];
  const mseq = [];

  sequence.map((seq, idx) => {
    if (idx === 0) {
      pseq.push(seq);
      mseq.push(-seq);
    } else if (idx % 2 === 0) {
      pseq.push(Math.max(pseq[idx - 1] + seq, seq));
      mseq.push(Math.max(mseq[idx - 1] - seq, -seq));
    } else {
      pseq.push(Math.max(pseq[idx - 1] - seq, -seq));
      mseq.push(Math.max(mseq[idx - 1] + seq, seq));
    }

    answer = Math.max(answer, pseq[idx], mseq[idx]);
  });

  return answer;
}

File: 프로그래머스/3/161988. 연속 펄스 부분 수열의 합/연속 펄스 부분 수열의 합.py
def solution(sequence):
    first_plus_max = 0
    first_minus_max = 0
    answer = 0
    for i, n in enumerate(sequence):
        if i % 2 == 0:
            first_plus_max = max(first_plus_max+n, n)
            first_minus_max = max(first_minus_max-n, -n)
        else:
            first_plus_max = max(first_plus_max-n, -n)
            first_minus_max = max(first_minus_max+n, n)
        answer = max(answer, first_plus_max, first_minus_max)
    return answer